Provide stake pool desirability scores and pledge status

HeinrichApfelmus opened the pull request:
### Issue number

ADP-996, ADP-507

### Overview

In this PR, we add additional metrics to the stake pool listing in order to be able to display fairer pool rankings. Specifically, we add a *stake pool desirability score* and the *pledge status* to the `GET /v2/stake-pools` endpoint. This information is intended to supplement the ranking of pools whose non-myopic member rewards ranking is very low.

The *desirability score* reflects how cost-effective and reliable the pool operator runs their pool, and how much of their own stake they are willing to pledge, but does not take the current stake delegated to the pool into account. In the long run, only the first `k` pools with highest desirability scores are expected to attract delegators, where `k` is a protocol parameter (`k = 500` as of 13th Jan 2022). However, in the short term, rewards for delegators can be much lower than the desirability score, as the pool has not attracted enough stake to reach saturation; also, good estimates of the pool's block production performance may not be available yet.

The *pledge status* indicates whether a pool owner has delegated enough stake to their own pool in order to meet the pledge that they made when registering their pool. A pool where the `owner_stake_lower_than_pledge` flag is set will earn *no rewards*, which is a good reason to delegate elsewhere.

See also the [REST API documentation](https://input-output-hk.github.io/cardano-wallet/api/edge/#tag/Stake-Pools) for more information (available once this PR is merged).

### Details

* Example endpoint response with new fields:
```graphql
{
  …,
  "metrics": {
    …,
    "desirability_score": 0,
    "owner_stake": {
      "quantity": 0,
      "unit": "lovelace"
    },
  },
  "flags": [
    "owner_stake_lower_than_pledge"
  ],
}
```
* Computing the pool metrics takes a bit of time. In order to speed up subsequent queries, the wallet process caches the computation. This behavior can be turned off with the `--no-cache-listpools` command line flag. The cache time-to-live (TTL) can be modified with the `--cache-listpools-ttl DURATION` flag (default: 1 hour). See `cardano-wallet --help` for more information.

### Comments

* At the moment, the non-myopic rewards are used to rank stake pools in Daedalus, and the desirability score is intended to supplement this information. Research at IOG is currently looking into pool metrics that are easier to grasp and better reflect rewards in both the short and the long term.

Commit:
Add more ranking information to StakePoolSummary

Introduces and populates new fields in the GET /stake-pools endpoint.

- `metrics` > `desirability_score` provides an additional score for ranking pools with zero `non_myopic_member_rewards`
- `metrics` > `owner_stake` records the stake delegated by a stake pool owner in order to meet his/her pledge.
- The convenience flag `owner_stake_lower_than_pledge` highlights when a stake pool does not meet his/her pledge.

WARNING: The implementation in this commit may be too slow for mainnet.

Note: The `performanceEstimate` is not a `Percentage`, but a `Rational`. Pools may produce more blocks than expected due to the random nature of the block production lottery. In fact, if there is a pool that had bad luck in the lottery, then there must necessarily be at least one with pool with better than average luck.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -737,6 +740,28 @@ fromNonMyopicMemberRewards =
+    . Map.mapKeys fromPoolId
+    . SL.desirabilities
+    where
+    fromDesirability (SL.Desirability x y) = W.StakePoolDesirability x y
```
Hi @HeinrichApfelmus 

Just a tiny point of style! We usually deindent `where` by two spaces, to make it stand out visually. So like this:

```suggestion
getDesirabilities =
    Map.map fromDesirability
    . Map.mapKeys fromPoolId
    . SL.desirabilities
  where
    fromDesirability (SL.Desirability x y) = W.StakePoolDesirability x y
```

Commit:
Add simple in-memory cache for stake distribution

NOTE: The cache is not activated in this commit yet.

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -228,17 +229,18 @@ newStakePoolLayer
 newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
-    pure $ StakePoolLayer
+    (worker, _stakeDistribution) <- nooooCacheWorker hour (stakeDistribution nl) 
+    pure $ (worker, StakePoolLayer
```
HLint will probably whinge about this line.

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -277,48 +279,24 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
-                lsqData
-                dbData
+        mnlData <- stakeDistribution nl
+        case mnlData of
```
```suggestion
        stakeDistribution nl >>= \case
```

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -332,144 +310,252 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
+
+-- | Create a new cache and a worker to fill it.
+newCacheWorker
+    :: Int -- ^ time period in seconds
```
```suggestion
    :: NominalDiffTime -- ^ cache TTL
```

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -332,144 +310,252 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
+            STM.atomically $ writeCache cache a
+            threadDelay (max 0 (seconds-3) * 1_000_000)
+    return (CacheWorker $ worker, STM.atomically $ readCache cache)
+    -- TODO: make cache worker more intelligent
```
Does it need to be much more intelligent?
Perhaps the cache period could be defined in terms of slots/epochs.
This looks like it pre-fills the cache shortly after startup. I think that's good.
It should probably recover from exceptions in the action though (perhaps by retrying a few times).
And some debug-level tracing would be good in case anything goes wrong with the queries.

HeinrichApfelmus replied:
> It should probably recover from exceptions in the action though (perhaps by retrying a few times).
> And some debug-level tracing would be good in case anything goes wrong with the queries.

This is the intelligence I was looking for but couldn't think of myself. Thanks! 😊

rvl reviewed (commented):
Looks great!!

Before PR approval, we will just need some testing that the non-myopic rankings calculated by the wallet are the same as what would be calculated by the ledger. Currently I'm not sure what the best way to do this might be. Using the local cluster integration tests may be the easiest way.

I notice that the TVar caching is present but disabled.
Without caching, it would still be just as fast as before, right?

I tried testing it out locally and got some PastHorizonExceptions - not sure yet whether this is due to your changes, or because of the Alonzo update in master branch.

Commit:
Add Debug message for `GetRewardProvenance` query

Commit:
Add variant of `threadDelay` with `NominalDiffTime`

Commit:
Add `Control.Cache` for caching long running actions

Commit:
Add retries to local state query caching

Also use the new `Control.Cache` module.

Commit:
Add NO_CACHE_LOCALSTATEQUERY env variable for testing

* The `serveWallet` function gets a new argument of type `Maybe DebugConfig`
* The above environment variable is *disabled* outside of integration tests and the local cluster

Commit:
Fix cache, add CACHE_LOCALSTATEQUERY_TTL env var

* Actually cache stuff.
* The above environment variable is *disabled* outside of integration tests and the local cluster. In a production environment, the default cache TTL is 60 minutes.

Commit:
Change treatment of new pool registration certificates

* We cannot say for certain yet that the owner has allocated the stake to back their pledge
* We cannot say for certain yet that the pool has produced blocks

These assumptions will affect the desirability scores.

Commit:
Update integration tests for stake pool metrics

* Add test for stake pool desirability.

* The non-myopic rewards are usually non-zero even if the user-submitted stake is zero. I have removed the integration test.

* It turns out that the total stake of the test cluster used for integration is so large that the given user stakes only make a minuscule difference. Our new non-myopic reward calculation uses `Double` internally for reasons of speed, which causes the "non-myopic rewards are based on stake" to fail due to limited precision. However, by changing the optimum number of pools from 3 to 2, we can ensure that one pool will lose out on desirability and its non-myopic rewards will change markedly with increasing user stake.

The notion of non-myopic rewards will be phased out, hence we do not need too pedantic about these integration tests.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -1070,42 +1081,11 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
                 rewards pool1 .> rewards pool3
 
-        it "non-myopic-rewards are based on stake" $ \ctx -> runResourceT $ do
-            eventually "rewards are smaller for smaller stakes" $ do
```
I'm wondering if it is not due to the caching perhaps? As far as I see on mainnet, both `desirability_score` and `non_myopic_member_rewards` do not change no matter what value of `stake` parameter is provided to the endpoint.

HeinrichApfelmus replied:
Can you try really big stakes on mainnet? 🤔 The `non_myopic_member_rewards` should change, especially on pools which have a low `non_myopic_member_rewards` score, when you supply a stake of `10000000000000000` (10 billion Ada).

piotr-iohk replied:
No, I don't see them changing even for 45 billion... e.g.

```
$ curl -X GET http://localhost:8091/v2/stake-pools?stake=1 | jq | grep -B 50 AGG1

[
  {
    "flags": [],
    "metrics": {
      "saturation": 0.0003302915840810417,
      "desirability_score": 2153289257678,
      "non_myopic_member_rewards": {
        "quantity": 2153289257678, <---------------------------------
        "unit": "lovelace"
      },
      "produced_blocks": {
        "quantity": 1,
        "unit": "block"
      },
      "relative_stake": {
        "quantity": 0,
        "unit": "percent"
      },
      "owner_stake": {
        "quantity": 15206293156,
        "unit": "lovelace"
      }
    },
    "cost": {
      "quantity": 340000000,
      "unit": "lovelace"
    },
    "margin": {
      "quantity": 0,
      "unit": "percent"
    },
    "pledge": {
      "quantity": 15200000000,
      "unit": "lovelace"
    },
    "metadata": {
      "homepage": "https://www.adagogo1.com",
      "name": "ADAgogo1",
      "ticker": "AGG1",

$ curl -X GET http://localhost:8091/v2/stake-pools?stake=45000000000000000 | jq | grep -B 50 AGG1

[
  {
    "flags": [],
    "metrics": {
      "saturation": 0.0003302915840810417,
      "desirability_score": 2153289257678,
      "non_myopic_member_rewards": {
        "quantity": 2153289257678, <---------------------------------
        "unit": "lovelace"
      },
      "produced_blocks": {
        "quantity": 1,
        "unit": "block"
      },
      "relative_stake": {
        "quantity": 0,
        "unit": "percent"
      },
      "owner_stake": {
        "quantity": 15206293156,
        "unit": "lovelace"
      }
    },
    "cost": {
      "quantity": 340000000,
      "unit": "lovelace"
    },
    "margin": {
      "quantity": 0,
      "unit": "percent"
    },
    "pledge": {
      "quantity": 15200000000,
      "unit": "lovelace"
    },
    "metadata": {
      "homepage": "https://www.adagogo1.com",
      "name": "ADAgogo1",
      "ticker": "AGG1",
```

piotr-iohk replied:
Ok, perhaps for smaller `non_myopic_member_rewards` they do change indeed:
```
$ curl -X GET http://localhost:8091/v2/stake-pools?stake=1 | jq | grep -B 35 GOAT4

    "metrics": {
      "saturation": 0.016319055947771045,
      "desirability_score": 42371242529,
      "non_myopic_member_rewards": {
        "quantity": 357659983, <---------------------------------
        "unit": "lovelace"
      },
      "produced_blocks": {
        "quantity": 44,
        "unit": "block"
      },
      "relative_stake": {
        "quantity": 0,
        "unit": "percent"
      },
      "owner_stake": {
        "quantity": 1075490098497,
        "unit": "lovelace"
      }
    },
    "cost": {
      "quantity": 340000000,
      "unit": "lovelace"
    },
    "margin": {
      "quantity": 2.5,
      "unit": "percent"
    },
    "pledge": {
      "quantity": 1000000000000,
      "unit": "lovelace"
    },
    "metadata": {
      "homepage": "https://goatstake.com",
      "name": "Goat Stake",
      "ticker": "GOAT4",

$ curl -X GET http://localhost:8091/v2/stake-pools?stake=45000000000000000 | jq | grep -B 35 GOAT4

    "metrics": {
      "saturation": 0.016319055947771045,
      "desirability_score": 42371242529,
      "non_myopic_member_rewards": {
        "quantity": 42371242529, <---------------------------------
        "unit": "lovelace"
      },
      "produced_blocks": {
        "quantity": 44,
        "unit": "block"
      },
      "relative_stake": {
        "quantity": 0,
        "unit": "percent"
      },
      "owner_stake": {
        "quantity": 1075490098497,
        "unit": "lovelace"
      }
    },
    "cost": {
      "quantity": 340000000,
      "unit": "lovelace"
    },
    "margin": {
      "quantity": 2.5,
      "unit": "percent"
    },
    "pledge": {
      "quantity": 1000000000000,
      "unit": "lovelace"
    },
    "metadata": {
      "homepage": "https://goatstake.com",
      "name": "Goat Stake",
      "ticker": "GOAT4",
```

HeinrichApfelmus replied:
Yup. In fact, the definition of `sigma'` in the computation of `optimalRewards`,

https://github.com/input-output-hk/cardano-wallet/blob/85be73804bb6a243a8a8f73e13d144320c17dc46/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L364-L384

implies that we should actually *expect* the non-myopic rewards to *not* change with the user-supplied stake for the `nOpt` pools that rank best on desirability score — the non-myopic reward prediction assumes that they will each attract an equal share `z0 = 1/nOpt` of the stake, and adding more stake to them will *not* increase the rewards at all, as `sigma'` will not increase beyond `z0`.

If we want to replicate the concept of "smaller pool" in the integration test, I think we should set the optimum number of pools to `2` in the genesis file (or any number that is small than the total number of pools that stays active in the local cluster.)

piotr-iohk replied:
Ok, I understand that this change was introduced in this PR to comply with delegation design spec? 

>If we want to replicate the concept of "smaller pool" in the integration test, I think we should set the optimum number of pools to 2 in the genesis file (or any number that is small than the total number of pools that stays active in the local cluster.)

I think that's a good idea + changing this test to verify this. There are a lot of "smaller pools" on mainnet. So if we had nOpt = 2 (currently it is 3 I believe) then we would have, at least 1 "smaller pool" (as there are 3 pools in the integration tests). So, for this one pool the change of non-myopic rewards score should be visible in case of changing stake.

Commit:
Change data source for pool cost and margin

It turns out that we cannot rely on current pool certificates that fetched from the chain — we have to use a local state query to fetch the "current" pool parameters, which contain only changes made in the *previous* epoch. Detailed explanation in the new note [RewardEpoch].

Commit:
Fix deprecation warning

Use `import Cardano.Protocol.TPraos`

Commit:
Add `--cache-listpools-ttl` command line argument

* Also add `--no-cache-listpools`
* Rename the environment variables to `CACHE_LISTPOOLS_TTL` and `NO_CACHE_LISTPOOLS`.

Commit:
Add cache invalidation to `Control.Cache`

Not exported in the REST API yet.

HeinrichApfelmus committed:
Use the new `GetRewardInfoPools`

… and drop the other queries related to rewards.

This depends on cardano-node` version 1.33

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1294,8 +1310,10 @@ fromShelleyAddress = W.Address
 fromShelleyCoin :: SL.Coin -> W.Coin
 fromShelleyCoin (SL.Coin c) = Coin.unsafeFromIntegral c
 
+{-
```
Just delete it

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -328,144 +309,280 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
+    -> Coin -- ^ Stake that the user wants to delegate
+    -> Map poolId (PoolScore, RewardInfoPool, a)
+scorePools params pools t
+    = Map.fromList $ zipWith doScore sortedByDesirability areTop
```
We should first sort by non-myopic rewards, then sort by desirability, and then sort by seeded random order.

HeinrichApfelmus replied:
Ok, I will adjust the value to sort by. (The actual sorting is done in the `sortByReward` function, which currently sorts according to non-myopic rewards. The code snippet above is internal to the `scorePools` function.)

HeinrichApfelmus replied:
Actually, it turns out that the non-myopic member rewards will never be equal to exactly zero, as they are essentially equal to the rewards for pool stake = owner stake + user stake. Hence, there will probably never be any ties that are resolved by the additional sorting by desirability score. Nonetheless, I will add that.

rvl commented on lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml:
```diff
@@ -31,7 +31,7 @@ protocolParams:
   keyDeposit: 1000000
   keyDecayRate: 0
-  nOpt: 3
+  nOpt: 2
```
I think it's pretty hard to use the local cluster and integration tests for testing this change.

Testing the new pool listing on existing mainnet pools and comparing with the old implementation will be more fruitful.

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -1764,42 +1764,74 @@ x-stakePoolSaturation: &stakePoolSaturation
   description: |
-    Saturation-level of the pool based on the desired number of pools aimed by the network.
-    A value above `1` indicates that the pool is saturated.
+    The saturation of a pool is the ratio of the current `relative_stake`
```
Nice updates to the docs.

rvl reviewed (commented):
Looks great, but I just have a query about the sorting of the pools list.

Would you also be able to write some explanation in the PR description of what will change for Daedalus users in the stake pools list? Because this is something we will want to highlight in the release notes of the next version.

Commit:
Refine sorting order

First `nonMyopicMemberRewards`, then `desirabilityScore`, then randomness from seed

rvl reviewed (approved):
:+1:

Commit:
Fix scaling of Non-Myopic Pool Member Rewards

Commit:
Fix Non-Myopic Pool Member Rewards: Use pledge instead of live stake

Commit:
Fix rounding: Use floor instead of round

Commit:
Rename --cache-listpools-ttl to --cache-listpools-refresh

and clarify its function
